src/geo/jts.clj

```diff
@@ -40,71 +143,206 @@
 
 (defn segment-at-idx
   "LineSegment from a LineString's point at index to index + 1."
-  [^org.locationtech.jts.geom.LineString linestring idx]
-  (org.locationtech.jts.geom.LineSegment. (coord (point-n linestring idx))
-                                          (coord (point-n linestring (inc idx)))))
+  [^LineString linestring idx]
+  (LineSegment. (coord (point-n linestring idx))
+                (coord (point-n linestring (inc idx)))))
+
 (defn linear-ring
-  "Given a list of Coordinates, creates a LinearRing."
-  [coordinates]
-  (.createLinearRing gf (into-array Coordinate coordinates)))
+  "Given a list of Coordinates, creates a LinearRing. Allows an optional SRID argument at end."
+  ([coordinates]
+   (.createLinearRing gf-wgs84 (into-array Coordinate coordinates)))
+  ([coordinates srid]
+   (.createLinearRing (gf srid) (into-array Coordinate coordinates))))
 
 (defn linear-ring-wkt
   "Makes a LinearRing from a WKT-style data structure: a flat sequence of
-  coordinate pairs, e.g. [0 0, 1 0, 0 2, 0 0]"
-  [coordinates]
-  (-> coordinates wkt->coords-array linear-ring))
+  coordinate pairs, e.g. [0 0, 1 0, 0 2, 0 0]. Allows an optional SRID argument at end."
+  ([coordinates]
+   (-> coordinates wkt->coords-array linear-ring))
+  ([coordinates srid]
+   (-> coordinates wkt->coords-array (linear-ring srid))))
 
 (defn polygon
   "Given a LinearRing shell, and a list of LinearRing holes, generates a
   polygon."
   ([shell]
    (polygon shell nil))
   ([shell holes]
-   (.createPolygon gf shell (into-array LinearRing holes))))
+   (.createPolygon (get-factory shell) shell (into-array LinearRing holes))))
 
 (defn polygon-wkt
   "Generates a polygon from a WKT-style data structure: a sequence of
   [outer-ring hole1 hole2 ...], where outer-ring and each hole is a flat list
   of coordinate pairs, e.g.
 
   [[0 0 10 0 10 10 0 0]
-   [1 1  9 1  9  9 1 1]]"
-  [rings]
-  (let [rings (map linear-ring-wkt rings)]
-    (polygon (first rings) (rest rings))))
+   [1 1  9 1  9  9 1 1]].
+
+   Allows an optional SRID argument at end."
+  ([rings]
+   (let [rings (map linear-ring-wkt rings)]
+     (polygon (first rings) (into-array LinearRing (rest rings)))))
+  ([rings srid]
+   (let [rings (map #(linear-ring-wkt % srid) rings)]
+     (polygon (first rings) (into-array LinearRing (rest rings))))))
 
 (defn multi-polygon
   "Given a list of polygons, generates a MultiPolygon."
   [polygons]
-  (.createMultiPolygon gf (into-array Polygon polygons)))
+  (.createMultiPolygon (get-factory (first polygons)) (into-array Polygon polygons)))
 
 (defn multi-polygon-wkt
   "Creates a MultiPolygon from a WKT-style data structure, e.g. [[[0 0 1 0 2 2
-  0 0]] [5 5 10 10 6 2]]"
-  [wkt]
-  (multi-polygon (map polygon-wkt wkt)))
+  0 0]] [5 5 10 10 6 2]]. Allows an optional SRID argument at end."
+  ([wkt]
+   (multi-polygon (map polygon-wkt wkt)))
+  ([wkt srid]
+   (multi-polygon (map #(polygon-wkt % srid) wkt))))
+
+(defn- coord-dimension-check
```

Comment:
I'd love to talk a little more about the various 3-D checks in these functions. I'm curious if this is something that's required in order to be compatible/complete with proj4j or if we just think we should have it for completeness of the API.

These changes are already getting quite large and checking for 2d/3d support everywhere adds a lot of complexity. This library has never purported to support 3d operations before, so unless we are trapped into this by one of the dependencies I am starting to think I'd be happy to remove all the 3d stuff, at least for now. That would simplify this batch of changes quite a bit, and if we think it's really a big improvement we could always consider adding it as a separate round of changes.

Reply:
I also have never needed 3d in my datasets. I don't think that this library needs to go out of its way to enable the creation of 3d shapes. If we drop all 3d support and stick to 2d for now, it wouldn't hurt my current use cases at all.

However, jts, ewkb, and geojson very much do support 3d. Proj4j isn't very good at 3d transformations, but if it gets a geometry with a lat, long, and 0 z dimension, it'll transform into a geometry with an x, y, and 0 z dimension. (In fact, projections are about to start thinking much more in earnest [about four dimensions](https://gdalbarn.com/), to account for changes in 3 dimensions over time!)

If we were using the full parent proj.4 instead of its pure java port, the 3d functionality in projections would be much more seamless. There is a JNI wrapper around proj.4, but using that would require that this library start dealing with keeping compiled versions of proj.4 for different platforms and integrating that, so I do think that sticking to proj4j is the right move. As proj4j starts to better handle 3d coordinates, I think that the loss of a third dimension becomes more noticeable.

Since the library can read formats in and its dependencies will create things with 3 dimensions, and since the dependencies can take 3 dimensional things and write them out correctly to the various formats, the one place where the library's core functionality may want to be 3d aware is where it interacts with coordinates and transformations. If we don't at least pass the z through the coordinate, any transformation will drop the z dimension on its way through the library. If 3d is removed from the library, we'll need to clarify that as a caveat in the io namespace.